packages/apps/plugins/plugin-stack/src/components/StackMain.tsx

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import React, { Fragment, useEffect, useState } from 'react';
+
+import { Main, Input, List, Button, useTranslation, randomString } from '@dxos/aurora';
+import { defaultBlockSeparator, mx } from '@dxos/aurora-theme';
+import { subscribe } from '@dxos/observable-object';
+import { Surface } from '@dxos/react-surface';
+
+import { StackModel, StackProperties, StackSectionModel, StackSections } from '../props';
+
+// todo(thure): `observer` causes infinite rerenders if used here.
+const StackMainImpl = ({ sections }: { sections: StackSections }) => {
+  const { t } = useTranslation('dxos:stack');
+  const [_, setIter] = useState([]);
+  useEffect(() => {
+    // todo(thure): TypeScript seems to get the wrong return value from `ObservableArray.subscribe`
+    return sections[subscribe](() => setIter([])) as () => void;
+  }, []);
+  return (
+    <article>
+      {sections
+        // todo(thure): This filter should be unnecessary; why is the first (or only?) value sometimes some sort of array-like object?
+        .filter((section) => 'source' in section)
+        .map((section, o) => {
+          return (
+            <Fragment key={section.source.guid}>
+              <Surface role='section' data={section} />
+            </Fragment>
+          );
+        })}
+      <div role='none' className='p-4'>
+        <Button
+          onClick={() => {
+            const section: StackSectionModel = {
+              source: { resolver: 'dxos:markdown', guid: randomString() },
+              object: {
+                id: randomString(),
+                content: '',
+                title: '',
+              },
+            };
+            sections.splice(sections.length, 0, section);
+          }}
+        >
+          {t('add section label')}
+        </Button>
+      </div>
+    </article>
+  );
+};
+
+export const StackMain = ({
+  data: [stack, properties],
+}: {
+  data: [stack: StackModel, properties: StackProperties];
+}) => {
+  const { t } = useTranslation('dxos:stack');
+  return (
+    <Main.Content classNames='min-bs-[100vh] mli-auto max-is-[60rem] bg-white dark:bg-neutral-925'>
+      <Input.Root>
```

Comment:
this looks right, although we may want a lower-level reusable component for header (and other things that might live here) shared with the composer plugin. and similarly we may want an low-level abstraction of the stack itself that can be reused in the Presenter plugin which might not have a title/header. so just wanted to plant this thought before the concrete starts to dry.

Reply:
Sure, I agree. This logic is contained by a component called `StackMain`, ‘main’ indicating its role, so the idea is this is for a specific use-case and may have patterns specific to the role that can be factored out (like the input), or patterns all Stacks should use that can be factored out.